### PR TITLE
backend:jottacloud change api used by ListR ( --fast-list )

### PR DESCRIPTION
**Pull request by Kimbsen**
#### What is the purpose of this change?

To update the jottacloud backend to use a better performing api for the ListR() call. The current api `mode=list` has poor time to first byte and poor support for very large trees. The `mode=liststream` is better in both of those. 

#### Was the change discussed in an issue or in the forum before?

No.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by albertony** [comment]
Looking back at my notes - though 3 years old 😪... I considered `mode=liststream`, but found it did only work for requests at root of mountpoint, i.e. could not be used to list from any given folder. I guess this has been fixed now?

**Comment by Kimbsen** [comment]
@albertony It has been fixed yes.

**Comment by albertony** [comment]
Great!

Had a very quick look at the code (I'll leave the full review up to others):
- Seems it no longer takes the paths through `f.opt.Enc.ToStandardName`, if I'm not mistaken, you need that to correctly handle filenames with special characters (https://rclone.org/jottacloud/#restricted-filename-characters).
- Also, I notice this version does not get the mime type information from backend, and just hard codes it to `application/octet-stream`. That may be a problem in some cases, but perhaps acceptable as a documented limitation, I don't know.

**Comment by Kimbsen** [comment]
The mimetype cannot be provided by this api no. But i'll look into using `f.opt.Enc.ToStandardName`.

**Comment by ncw** [comment]
Do the integration tests (which test recursive listing) work OK?

    go install ./...
    test_all -backends jottacloud

You also need to run gofmt on the files I think according to the linter.

**Comment by buengese** [comment]
If I recall this correctly, we do not guarantee that `Object.MimeType` returns a mime type even for backends that support this feature, so you should probably leave mime type empty instead of hard coding it.

As a side note: your test case for `TestParseListRStream` feels unnecessary large to me. It's not big issue but I'd like to avoid clutter

**Comment by albertony** [comment]
> If I recall this correctly, we do not guarantee that `Object.MimeType` returns a mime type even for backends that support this feature, so you should probably leave mime type empty instead of hard coding it.

True:

https://github.com/rclone/rclone/blob/38dc3e93eef9cd38f0668ce7efca73a4ca188cbe/fs/types.go#L149-L151

**Comment by Kimbsen** [comment]
> 
> 
> Do the integration tests (which test recursive listing) work OK?
> 
> ```
> go install ./...
> test_all -backends jottacloud
> ```
> 
> You also need to run gofmt on the files I think according to the linter.

The tests were failing. They are passing now though after https://github.com/rclone/rclone/pull/5823/commits/317f09aba203e3696e7fad98d4635da1f22f265d. Thanks for pointing those out to me. Kind of embarrased to not have run them before submitting the initial pr :) Diving into why they were failing was very informative :)


> 
> 
> If I recall this correctly, we do not guarantee that `Object.MimeType` returns a mime type even for backends that support this feature, so you should probably leave mime type empty instead of hard coding it.
> 
> As a side note: your test case for `TestParseListRStream` feels unnecessary large to me. It's not a big issue but I'd like to avoid clutter

I've now left the mimetype blank and i removed the test. The integration tests tests for the same thing so it was redundant.

The commit history is abit messy now though. Should i squash it?

**Comment by ncw** [comment]
> > Do the integration tests (which test recursive listing) work OK?
> The tests were failing. They are passing now though after [317f09a]

Great!

> I've now left the mimetype blank and i removed the test. The integration tests tests for the same thing so it was redundant.

Sound good.

> The commit history is abit messy now though. Should i squash it?

Yes, please rebase on master, squash and force push.

@buengese do you want to give this a review when that is done?

**Comment by Kimbsen** [comment]
> Yes, please rebase on master, squash and force push.

Done :)

File: backend/jottacloud/jottacloud.go [hunk 1]
```diff
@@ -988,12 +1061,29 @@ func (f *Fs) ListR(ctx context.Context, dir string, callback fs.ListRCallback) (
...
+		resp, err = f.srv.Call(ctx, &opts)
+		if err != nil {
+			return shouldRetry(ctx, resp, err)
+		}
+
+		// raw path is /username/device/mountpoint/path
+		// liststream paths are /mountpoint/path
+		trimPrefix := f.filePathRaw("")
+		trimPrefix = trimPrefix[strings.Index(trimPrefix, "/")+1:]
+		trimPrefix = trimPrefix[strings.Index(trimPrefix, "/"):]
```
Comment (buengese) [inline backend/jottacloud/jottacloud.go:1078]: Instead of building the root path using filePathRaw and trimming stuff back of it would probably be cleaner to just build the path you need here.

File: backend/jottacloud/jottacloud.go [hunk 2]
```diff
@@ -931,49 +932,121 @@ func (f *Fs) List(ctx context.Context, dir string) (entries fs.DirEntries, err e
...
+			md5:         f.Checksum,
+			modTime:     time.Time(f.Modified),
+		})
+	}
+
+	trimPathPrefix := func(p string) string {
+		p = strings.TrimPrefix(p, path.Join(trimPrefix))
+		p = strings.TrimPrefix(p, "/")
+		return p
+	}
```
Comment (buengese) [inline backend/jottacloud/jottacloud.go:994]: Path handling is still a bit messy and hard to follow, with you once again trimming off more stuff here. Also using `path.Join` to trim off a `/` suffix is far from clear.

File: backend/jottacloud/jottacloud.go [hunk 3]
```diff
@@ -931,49 +932,121 @@ func (f *Fs) List(ctx context.Context, dir string) (entries fs.DirEntries, err e
...
-func (f *Fs) listFileDir(ctx context.Context, remoteStartPath string, startFolder *api.JottaFolder, fn listFileDirFn) error {
-	pathPrefix := "/" + f.filePathRaw("") // Non-escaped prefix of API paths to be cut off, to be left with the remote path including the remoteStartPath
-	pathPrefixLength := len(pathPrefix)
-	startPath := path.Join(pathPrefix, remoteStartPath) // Non-escaped API path up to and including remoteStartPath, to decide if it should be created as a new dir object
-	startPathLength := len(startPath)
-	for i := range startFolder.Folders {
-		folder := &startFolder.Folders[i]
-		if !f.validFolder(folder) {
-			return nil
+type listStreamTime time.Time
```
Comment (buengese) [inline backend/jottacloud/jottacloud.go:935]: Do we actually have a third timeformat for Jottacloud now?  😨

**Review by buengese — COMMENTED** [review]
A bit of nitpick on the path handling. Besides that it looks fine.

File: backend/jottacloud/jottacloud.go (hunk 1 above)
Comment (Kimbsen) [reply backend/jottacloud/jottacloud.go:1078]: I agree. Will do that.

File: backend/jottacloud/jottacloud.go (hunk 2 above)
Comment (Kimbsen) [reply backend/jottacloud/jottacloud.go:994]: Yeah, that `path.Join()` should not have been there. I forgot it. It should just have been `p = strings.TrimPrefix(p, trimPrefix)`. Will fix that.

**Comment by Kimbsen** [comment]
I pushed those changes @buengese . Let me know if i should rebase of master and squash again.

**Comment by buengese** [comment]
That would be nice.

**Comment by Kimbsen** [comment]
Done.

**Comment by albertony** [comment]
There should probably be a note in this section that mime is not supported when using --fast-list:

https://rclone.org/jottacloud/#fast-list

**Comment by buengese** [comment]
Oh, that's very true. I always forget about doc chances
